runtime/dart_isolate.cc

```diff
@@ -0,0 +1,685 @@
+// Copyright 2017 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/runtime/dart_isolate.h"
+
+#include <cstdlib>
+#include <tuple>
+
+#include "flutter/fml/trace_event.h"
+#include "flutter/lib/io/dart_io.h"
+#include "flutter/lib/ui/dart_runtime_hooks.h"
+#include "flutter/lib/ui/dart_ui.h"
+#include "flutter/runtime/dart_service_isolate.h"
+#include "flutter/runtime/dart_vm.h"
+#include "lib/fxl/files/path.h"
+#include "lib/tonic/converter/dart_converter.h"
+#include "lib/tonic/dart_class_library.h"
+#include "lib/tonic/dart_class_provider.h"
+#include "lib/tonic/dart_message_handler.h"
+#include "lib/tonic/dart_state.h"
+#include "lib/tonic/dart_sticky_error.h"
+#include "lib/tonic/file_loader/file_loader.h"
+#include "lib/tonic/scopes/dart_api_scope.h"
+#include "lib/tonic/scopes/dart_isolate_scope.h"
+#include "third_party/dart/runtime/include/dart_tools_api.h"
+
+namespace blink {
+
+fml::WeakPtr<DartIsolate> DartIsolate::CreateRootIsolate(
+    const DartVM* vm,
+    TaskRunners task_runners,
+    std::unique_ptr<Window> window,
+    fml::WeakPtr<GrContext> resource_context,
+    fxl::RefPtr<flow::SkiaUnrefQueue> unref_queue,
+    std::string advisory_script_uri,
+    std::string advisory_script_entrypoint,
+    Dart_IsolateFlags* flags) {
+  TRACE_EVENT0("flutter", "DartIsolate::CreateRootIsolate");
+  Dart_Isolate vm_isolate = nullptr;
+  fml::WeakPtr<DartIsolate> embedder_isolate;
+
+  char* error = nullptr;
+
+  // Since this is the root isolate, we fake a parent embedder data object. We
+  // cannot use unique_ptr here because the destructor is private (since the
+  // isolate lifecycle is entirely managed by the VM).
+  DartIsolate* root_embedder_data =
+      new DartIsolate(vm,                           // VM
+                      task_runners,                 // task runners
+                      std::move(resource_context),  // resource context
+                      std::move(unref_queue),       // skia unref queue
+                      advisory_script_uri,          // advisory URI
+                      advisory_script_entrypoint    // advisory entrypoint
+      );
+
+  std::tie(vm_isolate, embedder_isolate) = CreateDartVMAndEmbedderObjectPair(
+      advisory_script_uri.c_str(),         // advisory script URI
+      advisory_script_entrypoint.c_str(),  // advisory script entrypoint
+      nullptr,                             // package root
+      nullptr,                             // package config
+      flags,                               // flags
+      root_embedder_data,                  // parent embedder data
+      &error                               // error (out)
+  );
+
+  delete root_embedder_data;
+
+  if (error != nullptr) {
+    free(error);
```

Comment:
These errors are already logged in `DartIsolate::CreateDartVMAndEmbedderObjectPair`. I log there instead because that method will also be called when attempting to launch non-root isolates.